### PR TITLE
Cleaned up CMT lib by refactoring away a couple common functions

### PR DESCRIPTION
**Pull request by jarry-xiao**
- Created a function called `traverse_to_parent` to abstract away the ugly (and highly duplicated) code that hashes a reference node up the tree
- Created a function inside the `ChangeLog` class that handles proof node or leaf replacement. If this existed months ago, we would have never introduced the bug in replacing the RML
- Use `empty_node_cached` when possible (O(N^2) -> O(N)) preprocessing
- Refactored `append` to be more easily readable
- Moved most direct assignment of variables into functions

All test cases pass and smoke test is chugging along

### CODE REVIEW

File: lib/concurrent-merkle-tree/src/utils.rs [hunk 1]
```diff
@@ -31,19 +25,23 @@ pub fn empty_node_cached<const N: usize>(level: u32, cache: &mut Box<[Node; N]>)
...
+    for (depth, sibling) in proof.iter().enumerate() {
+        traverse_to_parent(&mut current_node, sibling, index >> depth & 1 == 0);
     }
-
     current_node
 }
 
+/// Computes the parent node of `node` and `sibling` and copies the result into `node`
+#[inline(always)]
+pub fn traverse_to_parent(node: &mut Node, sibling: &Node, is_left: bool) {
```
Comment (ngundotra) [inline lib/concurrent-merkle-tree/src/utils.rs:36]: i think `hash_to_parent` is a better name

File: lib/concurrent-merkle-tree/src/state.rs [hunk 2]
```diff
@@ -23,27 +23,46 @@ impl<const MAX_DEPTH: usize> ChangeLog<MAX_DEPTH> {
...
-                start.copy_from_slice(res.as_ref());
-            } else {
-                let res = hashv(&[s.as_ref(), start.as_ref()]);
-                start.copy_from_slice(res.as_ref());
-            }
-            if ix < MAX_DEPTH - 1 {
-                self.path[ix + 1] = start;
-            }
+    pub fn recompute_path(&mut self, index: u32, mut node: Node, proof: &[Node]) -> Node {
+        self.index = index;
```
Comment (ngundotra) [inline lib/concurrent-merkle-tree/src/state.rs:41]: why does recomputing path set the `index` of the leaf?

File: lib/concurrent-merkle-tree/src/merkle_roll.rs [hunk 3]
```diff
@@ -395,43 +366,34 @@ impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> MerkleRoll<MAX_DEPTH,
...
-                    let critbit_index = (MAX_DEPTH - 1) - common_path_len;
-                    self.rightmost_proof.proof[critbit_index] = change_log.path[critbit_index];
-                } else {
-                    self.rightmost_proof.leaf = change_log.get_leaf();
-                }
+                change_log.update_proof_or_leaf(
+                    self.rightmost_proof.index - 1,
+                    &mut self.rightmost_proof.proof,
+                    &mut self.rightmost_proof.leaf,
+                );
```
Comment (ngundotra) [inline lib/concurrent-merkle-tree/src/merkle_roll.rs:396]: this is the most useful part of the PR 👍

File: lib/concurrent-merkle-tree/src/state.rs [hunk 4]
```diff
@@ -23,27 +23,46 @@ impl<const MAX_DEPTH: usize> ChangeLog<MAX_DEPTH> {
...
+        proof: &mut [Node; MAX_DEPTH],
+        leaf: &mut Node,
+    ) {
+        let padding: usize = 32 - MAX_DEPTH;
+        if leaf_index != self.index {
+            // This bit math is used to identify which node in the proof
+            // we need to swap for a corresponding node in a saved change log
+            let common_path_len = ((leaf_index ^ self.index) << padding).leading_zeros() as usize;
+            let critbit_index = (MAX_DEPTH - 1) - common_path_len;
+            proof[critbit_index] = self.path[critbit_index];
```
Comment (ngundotra) [inline lib/concurrent-merkle-tree/src/state.rs:62]: it would be nice to refactor this using normal modular addition 🙃

**Review by ngundotra — COMMENTED** [review]
This PR would be better if we could refactor the remaining bit math to normal modular arithmetic.

Also, what does `Moved most direct assignment of variables into functions` mean?

File: lib/concurrent-merkle-tree/src/state.rs (hunk 2 above)
Comment (jarry-xiao) [reply lib/concurrent-merkle-tree/src/state.rs:41]: you're updating the change log

File: lib/concurrent-merkle-tree/src/state.rs (hunk 4 above)
Comment (jarry-xiao) [reply lib/concurrent-merkle-tree/src/state.rs:62]: I don't think there's any loss in safety or even any loss to the end user to do things this way. Every mutable public function calls `check_bounds`

**Review by ngundotra — APPROVED** [review]

**Pull request merged**